dns_server.go

```diff
@@ -19,129 +19,106 @@ import (
 var dnsR *dnsRecords
 var dnsServer *dns.Server
 var dnsAddr string
+var dnsSuffix string
 
 type dnsRecords struct {
 	sync.RWMutex
-	l               *logrus.Logger
-	dnsMap4         map[string]netip.Addr
-	dnsMap6         map[string]netip.Addr
-	hostMap         *HostMap
-	myVpnAddrsTable *bart.Lite
+	l      *logrus.Logger
+	dnsMap map[dns.Question][]dns.RR
 }
 
-func newDnsRecords(l *logrus.Logger, cs *CertState, hostMap *HostMap) *dnsRecords {
+func newDnsRecords(l *logrus.Logger) *dnsRecords {
 	return &dnsRecords{
-		l:               l,
-		dnsMap4:         make(map[string]netip.Addr),
-		dnsMap6:         make(map[string]netip.Addr),
-		hostMap:         hostMap,
-		myVpnAddrsTable: cs.myVpnAddrsTable,
+		l:      l,
+		dnsMap: make(map[dns.Question][]dns.RR),
 	}
 }
 
-func (d *dnsRecords) Query(q uint16, data string) netip.Addr {
-	data = strings.ToLower(data)
-	d.RLock()
-	defer d.RUnlock()
-	switch q {
-	case dns.TypeA:
-		if r, ok := d.dnsMap4[data]; ok {
-			return r
-		}
-	case dns.TypeAAAA:
-		if r, ok := d.dnsMap6[data]; ok {
-			return r
+func (d *dnsRecords) AddA(name string, addresses []netip.Addr) {
+	d.Lock()
+	defer d.Unlock()
+	q := dns.Question{Name: name, Qclass: dns.ClassINET, Qtype: dns.TypeA}
+	d.dnsMap[q] = nil
+
+	for _, addr := range addresses {
+		if addr.Is4() {
+			qType := dns.TypeToString[q.Qtype]
+			rr, err := dns.NewRR(fmt.Sprintf("%s %s %s", name, qType, addr.String()))
+			if err == nil {
+				d.dnsMap[q] = append(d.dnsMap[q], rr)
+				d.l.Debugf("DNS record added %s", rr.String())
+			}
 		}
 	}
-
-	return netip.Addr{}
 }
 
-func (d *dnsRecords) QueryCert(data string) string {
-	ip, err := netip.ParseAddr(data[:len(data)-1])
-	if err != nil {
-		return ""
-	}
-
-	hostinfo := d.hostMap.QueryVpnAddr(ip)
-	if hostinfo == nil {
-		return ""
-	}
-
-	q := hostinfo.GetCert()
-	if q == nil {
-		return ""
-	}
+func (d *dnsRecords) AddAAAA(name string, addresses []netip.Addr) {
+	d.Lock()
```

Comment:
this is minor, but it would be nice to make these `Add*` functions private, and then obtain the lock only once, in the `Add()` function